From 1f6dc6fbf1a0741edb6635a1798d1ac14802f8eb Mon Sep 17 00:00:00 2001
From: Jim Fehlig <jfehlig@suse.com>
Date: Tue, 5 Jul 2022 11:54:34 -0600
Subject: libxl: set script field of libxl_device_disk

Add a hack to the libvirt libxl driver to set
libxl_device_disk->script when the disk configuration starts
with some well-known Xen external block scripts: dmmd, drbd,
and npiv.

Signed-off-by: Jim Fehlig <jfehlig@suse.com>
---
 src/libxl/libxl_conf.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

Index: libvirt-8.10.0/src/libxl/libxl_conf.c
===================================================================
--- libvirt-8.10.0.orig/src/libxl/libxl_conf.c
+++ libvirt-8.10.0/src/libxl/libxl_conf.c
@@ -946,6 +946,20 @@ libxlDiskSetDiscard(libxl_device_disk *x
 }
 
 static void
+libxlDiskSetScript(libxl_device_disk *x_disk, const char *disk_spec)
+{
+    if (disk_spec == NULL)
+        return;
+
+    if (STRPREFIX(disk_spec, "dmmd:"))
+        x_disk->script = g_strdup("block-dmmd");
+    else if (STRPREFIX(disk_spec, "drbd:"))
+        x_disk->script = g_strdup("block-drbd");
+    else if (STRPREFIX(disk_spec, "npiv:"))
+        x_disk->script = g_strdup("block-npiv");
+}
+
+static void
 libxlDiskSetCacheMode(libxl_device_disk *x_disk, int cachemode)
 {
     switch (cachemode) {
@@ -1088,6 +1102,7 @@ libxlMakeNetworkDiskSrc(virStorageSource
 int
 libxlMakeDisk(virDomainDiskDef *l_disk, libxl_device_disk *x_disk)
 {
+    const char *src = virDomainDiskGetSource(l_disk);
     const char *driver = virDomainDiskGetDriver(l_disk);
     int format = virDomainDiskGetFormat(l_disk);
     virStorageType actual_type = virStorageSourceGetActualType(l_disk->src);
@@ -1101,7 +1116,7 @@ libxlMakeDisk(virDomainDiskDef *l_disk,
         if (libxlMakeNetworkDiskSrc(l_disk->src, &x_disk->pdev_path) < 0)
             return -1;
     } else {
-        x_disk->pdev_path = g_strdup(virDomainDiskGetSource(l_disk));
+        x_disk->pdev_path = g_strdup(src);
     }
 
     x_disk->vdev = g_strdup(l_disk->dst);
@@ -1207,6 +1222,8 @@ libxlMakeDisk(virDomainDiskDef *l_disk,
     x_disk->is_cdrom = l_disk->device == VIR_DOMAIN_DISK_DEVICE_CDROM ? 1 : 0;
     libxlDiskSetDiscard(x_disk, l_disk->discard);
     libxlDiskSetCacheMode(x_disk, l_disk->cachemode);
+    libxlDiskSetScript(x_disk, src);
+
     /* An empty CDROM must have the empty format, otherwise libxl fails. */
     if (x_disk->is_cdrom && !x_disk->pdev_path)
         x_disk->format = LIBXL_DISK_FORMAT_EMPTY;
